### PR TITLE
fix maintain heading + cleanup specified heading

### PR DESCRIPTION
**Pull request by witcpalek**

### CODE REVIEW

File: src/main/cpp/chassis/headingStates/MaintainHeading.cpp [hunk 1]
```diff
@@ -28,29 +28,24 @@ MaintainHeading::MaintainHeading() : ISwerveDriveOrientation(ChassisOptionEnums:
...
-        double error = abs(chassis->GetPose().Rotation().Degrees().to<double>() - chassis->GetStoredHeading().to<double>());
-        if (error < 5.0)
-            correction = CalcHeadingCorrection(chassis->GetStoredHeading(), m_kPMaintainFine);
-        else
-            correction = CalcHeadingCorrection(chassis->GetStoredHeading(), m_kPMaintainCoarse);
+        auto correction = units::angular_velocity::degrees_per_second_t(0.0);
+
+        auto translatingOrStrafing = (abs(chassisMovement.chassisSpeeds.vx.to<double>()) +
+                                      abs(chassisMovement.chassisSpeeds.vy.to<double>())) > 0.0;
+        if (abs(chassisMovement.chassisSpeeds.omega.to<double>()) > 0.0 || !m_prevTranslatinOrStrafing)
```
Comment (Tan560) [inline src/main/cpp/chassis/headingStates/MaintainHeading.cpp:39]: I don't see m_prevTranslatinOrStrafing being set any where? What is the purpose of this?

File: src/main/cpp/chassis/headingStates/MaintainHeading.h [hunk 2]
```diff
@@ -28,4 +28,7 @@ class MaintainHeading : public ISwerveDriveOrientation
     ~MaintainHeading() = default;
 
     void UpdateChassisSpeeds(ChassisMovement &chassisMovement) override;
+
+private:
+    bool m_prevTranslatinOrStrafing;
```
Comment (Tan560) [inline src/main/cpp/chassis/headingStates/MaintainHeading.h:33]: Should we initialize it to false?

File: src/main/cpp/chassis/headingStates/SpecifiedHeading.cpp [hunk 3]
```diff
@@ -26,17 +26,29 @@ SpecifiedHeading::SpecifiedHeading() : ISwerveDriveOrientation(ChassisOptionEnum
...
     if (chassis != nullptr)
     {
-        chassis->SetStoredHeading(chassis->GetPose().Rotation().Degrees());
+        auto error = chassis->GetPose().Rotation().Degrees() - m_targetAngle;
+        auto slot = abs(error.value()) < m_fineCoarseAngle.value() ? m_fineSlot : m_coarseSlot;
+        auto correction = CalcHeadingCorrection(m_targetAngle, kPSpecifiedHeading[slot]);
+        chassisMovement.chassisSpeeds.omega += correction;
+        // Not sure if this should be done as its inconsistent with maintain, but face speaker had it
+        // I'm inclined to believe this is correct as we now have a desired heading we want to maintain
+        chassis->SetStoredHeading(m_targetAngle);
```
Comment (Tan560) [inline src/main/cpp/chassis/headingStates/SpecifiedHeading.cpp:47]: Yeah we need to store the heading or if you leave specified angel you will go to the last stored heading and whip around. The only thing I think we should do/clean up is remove specified heading/Angel and just use maintain. But that might be a off season thing. Think it is fine the way it is

File: src/main/cpp/chassis/headingStates/MaintainHeading.cpp (hunk 1 above)
Comment (witcpalek) [reply src/main/cpp/chassis/headingStates/MaintainHeading.cpp:39]: Inadvertent delete when cleaning up.  Put it back.  Thx

File: src/main/cpp/chassis/headingStates/SpecifiedHeading.cpp (hunk 3 above)
Comment (witcpalek) [reply src/main/cpp/chassis/headingStates/SpecifiedHeading.cpp:47]: deleted comment

File: src/main/cpp/chassis/headingStates/MaintainHeading.h (hunk 2 above)
Comment (witcpalek) [reply src/main/cpp/chassis/headingStates/MaintainHeading.h:33]: initialized it to valse

**Review by Tan560 — APPROVED** [review]

File: src/main/cpp/chassis/headingStates/MaintainHeading.cpp [hunk 4]
```diff
@@ -22,35 +23,44 @@
...
+        {
+            auto deltaAng = units::angular_velocity::degrees_per_second_t(chassisMovement.chassisSpeeds.omega) *
+                            units::time::millisecond_t(20.0);
+            chassis->SetStoredHeading(chassis->GetPose().Rotation().Degrees() + deltaAng);
+        }
+        else if (translatingOrStrafing)
+        {
+            auto error = chassis->GetPose().Rotation().Degrees() - chassis->GetStoredHeading();
+            auto slot = abs(error.value()) < m_fineCoarseAngle.value() ? m_fineSlot : m_coarseSlot;
+            //    correction = CalcHeadingCorrection(chassis->GetStoredHeading(), kPMaintain[slot]);
```
Comment (Tan560) [inline src/main/cpp/chassis/headingStates/MaintainHeading.cpp:52]: Should probably clean this code up to only use the one controller, but I am fine leaving it here for now. @witcpalek

File: src/main/cpp/chassis/headingStates/MaintainHeading.h [hunk 5]
```diff
@@ -28,4 +29,9 @@ class MaintainHeading : public ISwerveDriveOrientation
     ~MaintainHeading() = default;
 
     void UpdateChassisSpeeds(ChassisMovement &chassisMovement) override;
+
+private:
+    bool m_prevTranslatinOrStrafing = false;
+    frc::PIDController m_fineController{5.0, 0.5, 0.0};
+    frc::PIDController m_coarseController{5.0, 0.5, 0.0};
```
Comment (Tan560) [inline src/main/cpp/chassis/headingStates/MaintainHeading.h:36]: Should probably clean this code up to only use the one controller, but I am fine leaving it here for now. @witcpalek

**Review by Tan560 — APPROVED** [review]

**Comment by Tan560** [comment]
@muscatd75 or @nickfrisch, I was the last to make changes so I can't review. This is a lot of changes, but it was what we ran on both robots tonight. Please take a look and approve/request changes if needed

**Review by nickfrisch — APPROVED** [review]

File: src/main/cpp/chassis/headingStates/MaintainHeading.cpp (hunk 4 above)
Comment (witcpalek) [reply src/main/cpp/chassis/headingStates/MaintainHeading.cpp:52]: I'll do it tonight

File: src/main/cpp/chassis/headingStates/MaintainHeading.h (hunk 5 above)
Comment (witcpalek) [reply src/main/cpp/chassis/headingStates/MaintainHeading.h:36]: I'll do it tonight

**Pull request merged**